controller/src/main/java/io/pravega/controller/store/stream/Int96CounterImpl.java

```diff
@@ -0,0 +1,140 @@
+/**
+ * Copyright Pravega Authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package io.pravega.controller.store.stream;
+
+import com.google.common.annotations.VisibleForTesting;
+import io.pravega.common.concurrent.SequentialProcessor;
+import io.pravega.common.lang.AtomicInt96;
+import io.pravega.common.lang.Int96;
+import io.pravega.common.tracing.TagLogger;
+import lombok.AllArgsConstructor;
+import lombok.Data;
+import org.slf4j.LoggerFactory;
+
+import javax.annotation.concurrent.GuardedBy;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.CompletionException;
+import java.util.concurrent.Executor;
+import java.util.function.BiConsumer;
+import java.util.function.BiFunction;
+
+/**
+  * Int96CounterImpl implements getNextCounter using the Int96Counter interface.
+  */
+class Int96CounterImpl implements Int96Counter {
+    private static final TagLogger log = new TagLogger(LoggerFactory.getLogger(Int96CounterImpl.class));
+
+    @VisibleForTesting
+    final BiFunction<CounterInfo, BiConsumer<Int96, Int96>, CompletableFuture<Void>> function;
+
+    private final Object lock;
+    @GuardedBy("lock")
+    private final AtomicInt96 limit;
+    @GuardedBy("lock")
+    private final AtomicInt96 counter;
+    private final SequentialProcessor sequentialProcessor;
+
+
+    public Int96CounterImpl(final BiFunction<CounterInfo, BiConsumer<Int96, Int96>, CompletableFuture<Void>> function,
+                            Executor executor) {
+        this.lock = new Object();
+        this.counter = new AtomicInt96();
+        this.limit = new AtomicInt96();
+        this.function = function;
+        this.sequentialProcessor = new SequentialProcessor(executor);
+    }
+
+    @Override
+    public CompletableFuture<Int96> getNextCounter(OperationContext context) {
+        CompletableFuture<Int96> future;
+        Int96 next;
+        synchronized (lock) {
+            next = counter.incrementAndGet();
+        }
+        if (next.compareTo(limit.get()) > 0) {
```

Comment:
Same here with the lock. If you are using intelij there is an option for lock annotation checking you can turn on which will catch these and mark them in the ide.

Reply:
Done